docs/userguide/apps.rst

```diff
@@ -3,107 +3,100 @@
+        import numpy as np
+        return np.multiply(x, m) + b
+
+
+2. *Global variables are inaccessible*.
+   Functions should not use variables defined outside the function.
+   Likewise, do not assume that variables created inside the function are visible elsewhere.
 
-A Python app may also act upon files. In order to make Parsl aware of these files,
-they must be specified by using the ``inputs`` and/or ``outputs`` keyword arguments,
-as in following code snippet, which copies the contents of one file (``in.txt``) to another (``out.txt``).
 
 .. code-block:: python
 
-       @python_app
-       def echo(inputs=[], outputs=[]):
-           with open(inputs[0], 'r') as in_file, open(outputs[0], 'w') as out_file:
-               out_file.write(in_file.readline())
+    # BAD: Uses global variables
+    global_var = {'a': 0}
 
-       echo(inputs=[in.txt], outputs=[out.txt])
+    @python_app
+    def counter_func(string: str, character: str = 'a'):
+        global_var[character] += string.count(character)  # `global_var` will not be accessible
 
-Special Keyword Arguments
-^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Any Parsl app (a Python function decorated with the ``@python_app`` or ``@bash_app`` decorator) can use the following special reserved keyword arguments.
+    # GOOD
+    @python_app
+    def counter_func(string: str, character: str = 'a'):
+        return {character: string.count(character)}
 
-1. inputs: (list) This keyword argument defines a list of input :ref:`label-futures` or files. 
-   Parsl will wait for the results of any listed :ref:`label-futures` to be resolved before executing the app.
-   The ``inputs`` argument is useful both for passing files as arguments
-   and when one wishes to pass in an arbitrary number of futures at call time.
-2. outputs: (list) This keyword argument defines a list of files that
-   will be produced by the app. For each file thus listed, Parsl will create a future,
-   track the file, and ensure that it is correctly created. The future 
-   can then be passed to other apps as an input argument.
-3. walltime: (int) This keyword argument places a limit on the app's
-   runtime in seconds. If the walltime is exceed, Parsl will raise an `parsl.app.errors.AppTimeout` exception.
+    for ch, co in good_global('parsl', 'a').result().items():
+        global_var[ch] += co
+
+
+Functions from Modules
+++++++++++++++++++++++
 
-Serializing Functions from Libraries
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The above rules do not apply for functions defined in Python modules.
+Functions in installed modules are sent to workers differently than functions defined in a script.
 
-Parsl can create Apps directly from functions defined in Python modules.
-Supply the function as an argument to ``python_app`` rather than creating a new function which is decorated.
+Supply a module function as an argument to ``python_app`` rather than creating a new function which is decorated.
```

Comment:
I mean, there is technically nothing stopping someone from defining a new function. It's just silly.

Suggestion:

To transform an imported function into an App, you do not need to create a new function to decorate. Rather, you can do the following:

Reply:
Good call. I changed the wording to reflect that it isn't necessary to define an own function